// @fe/comps/FileTree/index.js
/**
 * @file: index
 * @author: Cuttle Cong
 * @date: 2017/12/10
 * @description:
 */
import React from 'react'
import Tree from 'react-ui-tree'
import 'react-ui-tree/dist/react-ui-tree.css'
import PropTypes from 'prop-types'
import cx from 'classnames'
import uncontrolled from '@fe/utils/hoc/uncontrolled'

import './style.less'

@uncontrolled('tree')
export default class FileTree extends React.PureComponent {
  static propTypes = {
    ...Tree.propTypes,
    renderNode: PropTypes.func
  }
  static defaultProps = {
    isNodeCollapsed: false,
    paddingLeft: 20,
    tree: {
      'module': 'react-ui-tree',
      'children': [
        {
          'module': 'dist',
          'collapsed': true,
          'children': [
            {
              'module': 'node.js',
              'leaf': true
            },
            {
              'module': 'react-ui-tree.css',
              'leaf': true
            },
            {
              'module': 'react-ui-tree.js',
              'leaf': true
            },
            {
              'module': 'tree.js',
              'leaf': true
            }
          ]
        },
        {
          'module': 'example',
          'children': [
            {
              'module': 'app.js',
              'leaf': true
            },
            {
              'module': 'app.less',
              'leaf': true
            },
            {
              'module': 'index.html',
              'leaf': true
            }
          ]
        },
        {
          'module': 'lib',
          'children': [
            {
              'module': 'node.js',
              'leaf': true
            },
            {
              'module': 'react-ui-tree.js',
              'leaf': true
            },
            {
              'module': 'react-ui-tree.less',
              'leaf': true
            },
            {
              'module': 'tree.js',
              'leaf': true
            }
          ]
        },
        {
          'module': '.gitignore',
          'leaf': true
        },
        {
          'module': 'index.js',
          'leaf': true
        },
        {
          'module': 'LICENSE',
          'leaf': true
        },
        {
          'module': 'Makefile',
          'leaf': true
        },
        {
          'module': 'package.json',
          'leaf': true
        },
        {
          'module': 'README.md',
          'leaf': true
        },
        {
          'module': 'webpack.config.js',
          'leaf': true
        }
      ]
    },
    renderNode: null
  }
  state = {
    active: null,
    tree: this.props.tree
  }

  onClickNode = node => {
    this.setState({
      active: node
    })
  }

  renderNode = node => {
    return (
      <span
        className={cx('node', {
          'is-active': node === this.state.active
        })}
        onClick={this.onClickNode.bind(null, node)}
      >
        {node.module}
      </span>
    )
  }

  render() {
    return (
      <div className="tree">
        <Tree
          {...this.props}
          tree={this.state.tree}
          renderNode={this.renderNode}
        />
      </div>
    )
  }
}
